_posts/2022-10-08-02-pypy.md

```diff
@@ -0,0 +1,120 @@
+    - This .pyc file then gets interpreted to machine language in PVM(Python Virtual Machine) and finally you get your output
+
+ - This process of implementation is **CPython**.
+ - **Python is a programming language and CPython is implementation of python.**
+
+<br><br>
+ ## CPython:
+
+ - CPython is the default byte-code interpreter of Python, which is written in C.
+ - When you install python from https://www.python.org/, default iterpreter is CPython.
+ - **CPython basically convert your code to bytecode first and then interpreted this bytecode to machine language and gives you output.**
+ - CPython is slow
+
+
+<br><br>
+ ## PyPy:
+
+ - As we now know, CPython is slow, to fill this gap, PyPy is used.
+ - PyPy is written in RPython(Restricted Python - a subset of python.)
+ - But what makes PyPy fast? It is JIT Compiler.
+ - **PyPy uses JIT or Just-in-Time compilation approach to compile the source code into native machine code instead of bytecode. Thereby making the code to run faster.**
+ - PyPy is a compiler and CPython is interpreter.
+ - PyPy workd faster for only pure python code.
+ - So the case where PyPy works best is when executing long-running programs where a significant fraction of the time is spent executing Python code.
+
+ - Features of PyPy:
+   - Speed (4.4 times faster than CPython on an average.)
+   - Memory usage
+   - Stackless
+
+ - Drawbacks of PyPy:
+   - if your code doesn't run for at least a few seconds, then the JIT compiler won't have enough time to warm up.
+   - If all the time is spent in run-time libraries (i.e. in C functions), and not actually running Python code, the JIT compiler will not help.
+   - Not suitable for data science libraries as it contains C functions and not pure python code.
+
+<br><br>
+
+## Installation of PyPy3:
+
+- Ubuntu:
+
+  > sudo add-apt-repository ppa:pypy/ppa
+
+  > sudo apt update
+
+  > sudo apt install pypy3
+
+- Mac:
+
+  > brew install pypy3
+
+- Windows:
+
+  > Download Windows Binary file from https://www.pypy.org/download.html
+
+<br>
+## Installation of Pip:
+
+ - Download the latest version of get-pip.py from https://bootstrap.pypa.io/.
+ - go to the folder where you download get-pip and run following cmd:
+   > pypy get-pip.py
+
+## Running python code using pypy:
+
+ - First write your python code.
+ - then run following command instead of running > python3 <file_name.py>
+   > pypy3 <file_name.py>
+
+<br><br>
+## Comparison between CPython and PyPy
+
+ |![CPython](/assets/images/2022-10-08-00-python3-time-taken.png) |
+ |:--:|
+ | *time taken by CPython* |
+
+
+ |![PyPy](/assets/images/2022-10-08-00-pypy3-time-taken.png) |
+ |:--:|
+ | *time taken by PyPy* |
+
+
+As we can see by above example, PyPy took only 5 seconds to complete and CPython took 2 minutes and 20 seconds to complete.
+So use PyPy for pure python code and use CPython if you have libraries which are in C and python language.
+
+
+References:
+- https://www.pypy.org/
```

Comment:
```suggestion
### <span style="text-decoration: underline"> CPython & PyPy Execution </span>

Let's check upon actual execution time of same piece of code in both Cpython & PyPy for getting an idea of its difference! <br>
For comparision, we will execute a long iterating `for` loop and record execution period using `time`.

 |![CPython](/assets/images/2022-10-08-00-python3-time-taken.png) |
 |:--:|
 | *time taken by CPython* |


 |![PyPy](/assets/images/2022-10-08-00-pypy3-time-taken.png) |
 |:--:|
 | *time taken by PyPy* |

 > As we can see by above example, PyPy took only **5 seconds** to complete while CPython took **2 minutes and 20 seconds** to complete execution of same loop.

### <span style="text-decoration: underline"> Conclusion </span>

No doubt, Python (CPython) is vastly used for ML/AI projects as most of its libraries have embedded C functions included. So in such case, PyPy won't serve its purpose well !<br>
But for usecase involving pure python code, it will be best to use [**PyPy**](https://www.pypy.org/) instead of **CPython** for better execution efficiency. <br> 

```